File: day_2/e.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define INPUT_FILE "input.txt"

// + Helpers

char *trim(char *s) {
    char *ptr;
    if (!s) {
        return NULL;
    }
    if (!*s) {
        return s;
    }
    for (ptr = s + strlen(s) - 1; (ptr >= s) && isspace(*ptr); --ptr);
    ptr[1] = '\0';
    return s;
}

char ** split_string(char * str) {
    char ** tokens = malloc(sizeof(*tokens));
    int init_size = strlen(str);
	char delim[] = " ";

	char *ptr = strtok(str, delim);

    size_t counter = 0;
	while (ptr != NULL) {
        tokens = realloc(tokens, (counter + 1) * sizeof(*tokens));
        tokens[counter] = ptr;
		ptr = strtok(NULL, delim);
        counter++;
	}

	return tokens;
}

/**
 * This method is not optimized
 */
char ** get_inputs(size_t *lines_no) {
    FILE *fp;

    if ((fp = fopen(INPUT_FILE, "r")) == NULL){
        printf("Error when opening file!");
        exit(1);
    }

    *lines_no = 0;

    char * line = NULL;
    size_t len = 0;

    char **inputs = malloc(sizeof(*inputs));

    while (getline(&line, &len, fp) != -1) {
        inputs = realloc(inputs, (*lines_no + 1) * sizeof(*inputs));
        inputs[*lines_no] = malloc(sizeof(char) * len);
        strcpy(inputs[*lines_no], trim(line));

        (*lines_no)++;
    }

    fclose(fp);

    return inputs;
}

// - Helpers

typedef struct {
    char * kind;
    int value;
} movement;

movement * parse_movements(char **inputs, size_t lines_no) {
    movement * movements = malloc(lines_no * sizeof(movement));

    for (int i = 0; i < lines_no; i++) {
        char ** tokens = split_string(inputs[i]);
        movements[i].kind = tokens[0];
        movements[i].value = atoi(tokens[1]);
    }

    return movements;
}

int exercise_1(movement * movements, size_t movements_count) {
    int depth = 0;
    int distance = 0;
    for (int i = 0; i < movements_count; i++) {
        if (strcmp(movements[i].kind, "forward") == 0) {
            distance += movements[i].value;
        } else if (strcmp(movements[i].kind, "up") == 0) {
            depth -= movements[i].value;
        } else if (strcmp(movements[i].kind, "down") == 0) {
            depth += movements[i].value;
        }
    }
    return depth * distance;
}

int exercise_2(movement * movements, size_t movements_count) {
    int depth = 0;
    int distance = 0;
    int aim = 0;
    for (int i = 0; i < movements_count; i++) {
        if (strcmp(movements[i].kind, "forward") == 0) {
            distance += movements[i].value;
            depth += aim * movements[i].value;
        } else if (strcmp(movements[i].kind, "up") == 0) {
            aim -= movements[i].value;
        } else if (strcmp(movements[i].kind, "down") == 0) {
            aim += movements[i].value;
        }
    }
    return depth * distance;
}

int main() {
    size_t lines_no = 0;
    char ** inputs = get_inputs(&lines_no);
    movement * movements = parse_movements(inputs, lines_no);

    printf("Exercise 1: %d\n", exercise_1(movements, lines_no));
    printf("Exercise 2: %d\n", exercise_2(movements, lines_no));
}

File: day_2/e.ml
let file_name = "input.txt"

let read_lines file_name : string list =
  let ic = open_in file_name in
  let try_read () =
    try Some (input_line ic) with End_of_file -> None in
  let rec loop acc = match try_read () with
    | Some s -> loop (s :: acc)
    | None -> close_in ic; List.rev acc in
  loop []

type movement = {
  kind: string;
  value: int
}

let () =
    let exercise_1 movements =
      let depth, distance = List.fold_left (fun acc v ->
        let depth, distance = acc in
        match v with
        | {kind = "forward"; value} -> (depth, distance + value)
        | {kind = "up"; value} ->  (depth - value, distance)
        | {kind = "down"; value} ->  (depth + value, distance)
        | _ -> failwith "ERROR"
      ) (0, 0) movements in
      depth * distance
    in

    let exercise_2 movements =
      let depth, distance, _ = List.fold_left (fun acc v ->
        let depth, distance, aim = acc in
        match v with
        | {kind = "forward"; value} -> (depth + aim * value, distance + value, aim)
        | {kind = "up"; value} ->  (depth, distance, aim - value)
        | {kind = "down"; value} ->  (depth, distance, aim + value)
        | _ -> failwith "ERROR"
      ) (0, 0, 0) movements in
      depth * distance
    in

    let lines = read_lines file_name in
    let parse_input s =
      match String.split_on_char ' ' s with
      | kind :: value :: _ ->
        { kind = kind;
          value = int_of_string value}
      | _ -> failwith "ERROR"
    in
    let movements = List.map parse_input lines in

    Printf.printf "Exercise 1: %d\n" (exercise_1 movements);
    Printf.printf "Exercise 2: %d\n" (exercise_2 movements);
